cnap/core/stream.py

```diff
@@ -0,0 +1,347 @@
+
+    def _dev_num(self) -> int:
+        """
+        Get the number of device node.
+
+        For example, it should be 0 for /dev/video0, and 1 for /dev/video1
+
+        Returns:
+            int: The number of device node.
+
+        Raises:
+            ValueError: if the stream path is None or is not a valid video device node.
+            FileNotFoundError: if the camera's device node not exists.
+        """
+        if self.pathname is None:
+            raise ValueError("Stream path name is None")
+        if not re.match(r'/dev/video\d+', self.pathname):
+            LOG.error("The pathname %s is not a valid video device node", self.pathname)
+            raise ValueError(f"The pathname {self.pathname} is not a valid video device node")
+        if not os.path.exists(self.pathname):
+            LOG.error("The camera's device node %s not exists", self.pathname)
+            raise FileNotFoundError(f"The camera's device node {self.pathname} not exists")
+        return int(re.search(r'\d+', self.pathname).group())
+
+    def verify(self) -> bool:
+        return True
+
+    def read_raw_frame(self) -> numpy.ndarray:
+        ret, raw = self._device_obj.read()
+        if ret:
+            return raw
+        return None
+
+    def open(self) -> bool:
+        try:
+            dev_num = self._dev_num()
+        except ValueError as e:
+            LOG.error(e)
+            return False
+        except FileNotFoundError as e:
+            LOG.error(e)
+            return False
+        self._device_obj = cv2.VideoCapture(dev_num)
+        if not self._device_obj.isOpened():
+            LOG.error("Fail to open the camera, number: %d", dev_num)
+            return False
+        self._device_obj.set(cv2.CAP_PROP_FRAME_WIDTH, self.raw_size[0])
+        self._device_obj.set(cv2.CAP_PROP_FRAME_HEIGHT, self.raw_size[1])
+        self._device_obj.set(cv2.CAP_PROP_FPS, self.raw_fps)
+        return True
+
+    def close(self) -> None:
+        if self._device_obj is not None:
+            self._device_obj.release()
+
+
+class FileSource(StreamProvider):
+    """
+    File source stream provider implementation for providing stream input from video file.
+    """
+
+    DEFAULT_TARGET_FPS = 25
+
+    def __init__(self, name: str, pathname: str="classroom.mp4"):
+        """
+        Initialize a FileSource object.
+
+        Args:
+            name: The name of file source stream provider.
+            pathname: The path name of file source stream provider.
+        """
+        StreamProvider.__init__(self, name, pathname)
+        self._file_db = None
+        self._file_object = None
+        self._file_path = None
+        self._frame_counter = 0
+        self._max_frame = 0
+
+    def verify(self) -> bool:
+        return True
+
+    def open(self) -> bool:
+        LOG.debug("Open file source: %s", self._file_path)
+        try:
+            self._file_path = self.file_db.get_file(self.pathname)
+        except FileNotFoundError as e:
```

Comment:
I think if we catch exception, then we may use log.exception to handle

Reply:
Thanks, fixed.